content/post/impute-missing-levels.Rmd

```diff
@@ -0,0 +1,200 @@
+```
+
+In order to better understand we can look at the results of every `r ref("PipeOp")` separately.
+
+We can manually trigger the `r ref("PipeOp")` to test the operator on our task:
+
+```{r}
+ext_task = imp_missind$train(list(task))[[1]]
+ext_task$data()
+```
+
+For `r ref("mlr_pipeops_imputehist", "imputehist")`, we can do the same:
+
+```{r}
+ext_task = imp_num$train(list(task))[[1]]
+ext_task$data()
+```
+
+This time we obtain the imputed data set without `missing` values.
+
+```{r}
+ext_task$missings()
+```
+
+## Putting everything together
+
+Now we have to put all `r ref("PipeOp")`s  together in order to form a graph that handles imputation automatically.
+
+We do this by creating a `r ref("Graph")` that copies the data twice, processes each copy using the respective imputation method and afterwards union the features.
+For this we need the following two `r ref("PipeOp")`s :
+* `r ref("mlr_pipeops_copy", "copy")`: Creates copies of the data.
+* `r ref("mlr_pipeops_featureunion", "featureunion")` Merges the two tasks together.
+
+
+```{r}
+graph = po("copy", 2) %>>% gunion(list(imp_missind, imp_num)) %>>% po("featureunion")
+```
+
+as a last step we append the learner we planned on using:
+
+```{r}
+graph = graph %>>% po(learner)
+```
+
+We can now visualize the resulting graph:
+
+```{r}
+graph$plot(html = TRUE)
+```
+
+## Resampling
+
+Correct imputation is especially important when applying imputation to held-out data during the `predict` step.
+If applied incorrectly, imputation could then leak info from the test set, which might invalidate our performance estimates.
+`r mlr_pkg("mlr3pipelines")` takes this complexity away from the user and handles correct imputation internally.
+
+By wrapping this graph into a  `r ref("GraphLearner")`, we can now train resample the full graph, here with a 3-fold cross validation:
+
+```{r}
+glearner = GraphLearner$new(graph)
+rr = resample(task, glearner, rsmp("cv", folds = 3))
+rr$aggregate()
+```
+
+## Missing values during prediction
+
+In some cases, we have missing values only in the data we want to predict on.
+In order to showcase this, we create a copy of the task with several more missing columns.
+
+```{r}
+dt = task$data()
+dt[1:10, "age"] = NA
+dt[30:70, "pedigree"] = NA
+task2 = TaskClassif$new("pima2", dt, target = "diabetes")
+```
+
+And now we learn on `task`, while trying to predict on `task2`.
+
+```{r}
+glearner$train(task)
+glearner$predict(task2)
+```
+
+## Missing factor features
+
+For `factor` features, the process works analogously.
+Instead of using `r ref("mlr_pipeops_imputehist", "imputehist")`, we can for example use `r ref("mlr_pipeops_imputenewlvl", "imputenewlvl")`.
+This will simply replace every `NA` in each factor variable with a new value `missing`.
+
+A full graph might the look like this:
+
+```{r}
+imp_fct = po("imputenewlvl")
+graph = po("copy", 2) %>>%
+  gunion(list(imp_missind, imp_num %>>% imp_fct)) %>>%
+  po("featureunion")
```

Comment:
Does this really work? Can you apply this graph to some task? It reads as if we can have NAs in factor columns of the test set even without NAs in the training set. But that means that during training ranger never has seen `missing`. How will it handle this new factor for prediction?